lock/lock.go

```diff
@@ -0,0 +1,155 @@
+package lock
+
+import (
+	"fmt"
+	"github.com/gruntwork-io/go-commons/retry"
+	"github.com/sirupsen/logrus"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/session"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/gruntwork-io/gruntwork-cli/errors"
+)
+
+type LockTimeoutExceeded struct {
+	LockTable  string
+	LockString string
+	Timeout    time.Duration
+}
+
+func (err LockTimeoutExceeded) Error() string {
+	return fmt.Sprintf("Timeout trying to acquire lock %s in table %s (timeout was %s)", err.LockString, err.LockTable, err.Timeout)
+}
+
+
+// We assume that the DynamoDB table will be created prior to using this functionality. 
+// NewAuthenticatedSession gets an AWS Session, checking that the user has credentials properly configured in their environment.
+func NewAuthenticatedSession(awsRegion string) (*session.Session, error) {
+	sess, err := session.NewSession(aws.NewConfig().WithRegion(awsRegion))
+	if err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	if _, err = sess.Config.Credentials.Get(); err != nil {
+		return nil, errors.WithStackTrace(err)
+	}
+
+	return sess, nil
+}
+
+// We assume that the DynamoDB table will be created prior to using the locking mechanism
+// NewDynamoDb returns an authenticated client object for accessing DynamoDb
+func NewDynamoDb(awsRegion string) (*dynamodb.DynamoDB, error) {
+	sess, err := NewAuthenticatedSession(awsRegion)
+	if err != nil {
+		return nil, err
+	}
+	dynamodbSvc := dynamodb.New(sess)
+	return dynamodbSvc, nil
+}
+
+// AcquireLock will attempt to acquire the lock defined by the provided lock string in the configured lock table for the
+// configured region.
```

Comment:
This needs more context: namely, that we are using a DynamoDB table for locking, and we are using it in such a way that this will work as a distributed lock (as opposed to, say, a local lock for multi threading).